### ISSUE TITLE
Lost sessions when utilizing Block-Wise transfers in CoAP

### ISSUE THREAD

**Issue by rpb5bnc** [obgm/libcoap#372]
Hi,

We have implemented a CoAP based server using libcoap.  It is running EST over CoAP in server mode (https://tools.ietf.org/html/draft-ietf-ace-coap-est-12).  Because the Certificate Signing Requests (CSRs) and returned certificates exceed the path MTUs in the network, CoAP block-wise transfers needed to be supported.  libcoap requires the library for the next protocol layer up to implement part of the CoAP block-wise transfer functionality, for example, it needs to accumulate the block fragments of incoming CSRs until all blocks have been received and hold onto the outgoing certificate responses until all blocks have been sent.  This requires the EST library to maintain state on each session while the request is being processed.  While definitely not ideal, we have it working.  However, when any of these sessions fails to terminate, let's say a client node stops communicating or there is a network failure somewhere between our server and a client node, this session does not terminate normally.  We see that the libcoap library certainly must also maintain state for each session, and it appears to address this scenario of lost sessions by freeing up this state information for the oldest session when a new session comes in and there's no more space for a new session.  This works for libcoap, but there does not appear to be any mechanism to tell the application layer (in our case the EST library) that this session is now officially removed from libcoap's perspective.  The result is that the higher layer protocol that needs to maintain state to help with the CoAP protocol doesn't know when to remove its equivalent state information for this session.  It would seem important to keep these two layers of code in sync in this regard.

What I'd like to know is, is there something I'm missing in this API for libcoap when in server mode that will allow the application layer to be notified when a session is being removed from the CoAP protocol layer?

Thanks very much,

**Comment by mrdeep1** [comment]
As you are using coaps:// you can register an event handler and then act upon the event COAP_EVENT_DTLS_CLOSED when the session "goes away".  See coap_handle(3) man page for further details.

That said, I had been thinking on how to better handle blocked transfers - moving much more logic into the libcoap layer (beyond coap_add_data_blocked_response()) and one of the things was the ability to just send the next block rather than providing the complete data by the application on each block read request.

A challenge with that is that ideally the SIZE2 option should be included in at least the first block response, as well as a consistent ETAG so that the client can detect issues.

As per https://tools.ietf.org/html/rfc7959#section-4
````
...
Implementation Notes:

o As a quality of implementation consideration, block-wise transfers
for which the total size considerably exceeds the size of one
block are expected to include size indications, whenever those can
be provided without undue effort (preferably with the first block
exchanged)
````
Another challenge is that the client will be asking for the next block of data which will have a specific offset into the data stream being transferred.  There will have to be an assumption that the provided stream of data is being appropriately handled by the application.  All this is messy.

**Comment by rpb5bnc** [comment]
Hi, 

I'm sorry, I was interrupted by a couple of other issues and now getting back to this today.

I registered a callback with,

    coap_register_event_handler(ctx->coap_ctx, est_coap_event_handler);

and I do see my callback function being invoked when a session is connected and when the session completes normally with a close_notify alert from the client.  However, when I establish a DTLS session with a client and that client goes quiet after establishing the session what I see is the libcoap library clearing the session structure for this session 300 seconds later.  In this situation I don't see my callback being called.  This is the issue that I'm facing.  That is, when DTLS sessions do not terminate normally the application layer is not notified and we don't get a chance to clean up the structure used to maintain that session, but the libcoap library does clean up its structure, and at that point we get out sync with each other.

I don't mean to mix different subjects together but the following is somewhat related to the issues I'm looking into.  While investigating the above issue I noticed that when I ran a test client that opened 2048 DTLS connections to libcoap in server mode it allowed all these connections to be established.  Would there happen to be a way to limit the number of DTLS sessions that are established at any moment in time? 

One thing I will add is that we are currently using rc3.  We have not yet moved to the official 4.2.0.  Is it possible that this has been addressed sometime after 4.2.0.rc3?

Thanks in advance and best regards,
Peter

**Comment by mrdeep1** [comment]
Hi Peter,

Without #333 in your code (which is post 4.2.0), you are likely to run into "lost" DTLS sessions.

It does appear that an event does not trigger if COAP_DEFAULT_SESSION_TIMEOUT expires on a session - this is not correct and I will get it fixed.

It is possible to set the (undocumented) `context` variable `max_idle_sessions` to a number, and this will limit the maximum number of "idle" sessions (processed in `coap_endpoint_get_session`()).

**Comment by rpb5bnc** [comment]
Excellent!!!  Thank you so very much for the quick response.

As for `max_idle_sessions`, I had seen this in `coap_endpoint_get_session()` early on but at that time had not noticed that the only references of it were, 
```
../include/coap2/net.h:214:  unsigned int max_idle_sessions;  /**< Maximum number of simultaneous unused sessions per endpoint. 0 means no maximum. */
coap_session.c:484:  if (endpoint->context->max_idle_sessions > 0 &&
coap_session.c:485:      num_idle >= endpoint->context->max_idle_sessions) {
```
with no way to set this variable.  At the time of my first post I had thought this was how idle sessions were freed up, just one at a time, each time `coap_endpoint_get_session()` was called.  But when I got back to performing more tests this week and hit the server with 1000s of sessions I noticed the other loop that walks the entire linked list freeing up any that were idle for more than 5 minutes.

I'm on my way now to look at #333.  In fact, I'm going to see if it might be easier to just move our project forward to the official 4.2.0.  I suspect some people will be requesting that this be done soon anyway.

Again, thanks for the quick investigation and response on this, especially on the missed callbacks on the lost sessions.

Best regards,
Pete

**Comment by mrdeep1** [comment]
PR #379 fixes the missing event callback when a (D)TLS session idles out.

Currently the only way to set `max_idle_sessions` is to update the context variable directly as things stand.

**Comment by rpb5bnc** [comment]
Well, that was fast!  I will pull this in as well.  May take a little bit now because I've got some adjusting I need to do on our side.

Thanks very much,
Pete

**Comment by mrdeep1** [comment]
I'm thinking of adding in a new function `coap_add_data_large_response()` which basically has the same parameters as `coap_add_data_blocked_response()`.  The difference is that libcoap would handle the responses for the subsequent block requests by the client.

What I am not sure about is how the data is kept while doing the blocked transfers - one way is the application does not change the data until the entire data has been transmitted (perhaps needed because application is using mmap() of a large file) - another way is for libcoap to take a copy of (the potentially very large) data and letting the application re-use the data memory immediately.

Either way, it may make sense to have 2 new events - `data-transmitted` and `data-transmit-failed`. 

Having `coap_add_data_large_response()` and `coap_add_data_large_request()` would make things simpler for any application proxying data between, say, HTTP and CoAP.  In your case, EST would still have to assemble all of the outgoing certificate response, but not then worry so much about the onward blocked sending.

I need to separately think about a server handler where libcoap re-assembles a blocked POST/PUT request in libcoap heap and presents it as a single block of data to a 'large' handler.

What do you think?

**Comment by mrdeep1** [comment]
#554 (not yet integrated) now supports large body handling in libcoap - removing the need for applications to handle individual next block requests and provides a mechanism to free off the body data once the transfer is complete.

**Comment by mrdeep1** [comment]
@rpb5bnc  #554 is now a part of the develop branch. It removes a lot of the requirements for the application to the RFC7959 work, which is now done by libcoap.  So, an application can present a large data item to be send, libcoap sends all the payloads of the body and callbacks the application when finished (or on failure) so the application can free off the data.  Likewise, on receipt of all the payloads of a body, the application is informed once with all of the data (COAP_BLOCK_SINGLE_BODY needs to be set).

You may be interested in the work being done in https://datatracker.ietf.org/doc/draft-ietf-core-new-block/ which provides a means where the payloads of a body can be sent using Non-Confirmable messages quickly and has recovery mechanisms.  There is libcoap code that does this if you are interested based on #554 and is just another configuration option.

Please checkout #554 and then close down this issue if it works as you need it to.

**Comment by obgm** [comment]
PR #554 has been merged into develop. Does this resolve your issue?

**Comment by obgm** [comment]
Closed off due to inactivity.
